### PR TITLE
Dst 259  standardized1d v2

### PR DESCRIPTION
**Pull request by vegard-solum-4ss**
Functionality for generating standardized 1-D wave spectra + possibility to add two WaveSpectrum objects.

User story:

https://4insight.atlassian.net/jira/software/c/projects/DST/boards/10?modal=detail&selectedIssue=DST-259

### CODE REVIEW

File: src/waveresponse/_standardized1d.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    def _C(self, *args):
+        hs, tp, q = args
+        omega_p = 2.0 * np.pi / tp
+        c = (4.0 * q + 1.0) * omega_p**4 / 4.0
+        return (1.0 / 4.0) * (c**q * hs**2) / gammafun(q)
+
+    def _d(self, *args):
+        _, tp, q = args
+        omega_p = 2.0 * np.pi / tp
+        return (4 * q + 1) * omega_p**4 / 4.0
```
Comment (ace-e4s) [inline src/waveresponse/_standardized1d.py:397]: Same here, why use *args. Shouldnt we be more explicit.

File: tests/test_standardized1d.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        args = (hs, tp, gamma, sigma_a, sigma_b)
+        b_out = spectrum._b(*args)
+
+        omega_p = 2.0 * np.pi / tp
+        sigma = spectrum._sigma(omega_p, sigma_a, sigma_b)
+        b_expect = np.exp(-0.5 * ((spectrum._freq - omega_p) / (sigma * omega_p)) ** 2)
+
+        np.testing.assert_array_almost_equal(b_out, b_expect)
+
+    def test_reference_case(self):
```
Comment (ace-e4s) [inline tests/test_standardized1d.py:371]: use fixture?

File: tests/test_standardized1d.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+        hs = 3.5
+        tp = 10.0
+        spectrum_internal = spectrum._spectrum(freq, hs, tp, q=2.0)
+
+        freq_out, spectrum_out = spectrum(3.5, 10.0, q=2.0, freq_hz=True)
+
+        np.testing.assert_array_equal(freq_out, freq / (2.0 * np.pi))
+        np.testing.assert_array_equal(spectrum_out, spectrum_internal * (2.0 * np.pi))
+
+    def test_reference_case(self):
```
Comment (ace-e4s) [inline tests/test_standardized1d.py:429]: use fixture?

File: docs/user_guide/standardized_spectra.rst [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+.. of the significant wave height, Hs, and the wave peak period, Tp.
+
+
+Modified Pierson-Moskowitz
+..........................
+The *modified Pierson-Moskowits* spectrum (also known as Bretschneider or ISSC) is given by:
+
+.. math::
+
+    S_{PM}(\omega) = \frac{5}{16}H_s^2\omega_p^2\omega^{-5} exp\left(-\frac{5}{4} \left( \frac{\omega_p}{\omega} \right)^4 \right)
```
Comment (ace-e4s) [inline docs/user_guide/standardized_spectra.rst:31]: omega_p^4 instead of omega_p^2?

**Review by ace-e4s — CHANGES_REQUESTED** [review]
See comments

File: docs/user_guide/standardized_spectra.rst (hunk 4 above)
Comment (vegard-solum-4ss) [reply docs/user_guide/standardized_spectra.rst:31]: Yes, omega_p^4 is correct

File: src/waveresponse/_standardized1d.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+        Generate wave spectrum.
+
+        Parameters
+        ----------
+        *args
+            Spectrum parameters.
+        freq_hz : bool, optional
+            Whether to return the frequencies and spectrum in terms of rad/s (`True`)
+            or Hz (`False`). If `None` (default), the original units of `freq` is
+            preserved.
```
[the remark this answers is not in the document]
Comment (vegard-solum-4ss) [reply src/waveresponse/_standardized1d.py:30]: Yes, this docstring is wrong. Must be fixed.

File: src/waveresponse/_standardized1d.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+    and,
+
+    - ``S_pm(w)`` is the Pierson-Moskowitz (PM) spectrum.
+    - ``gamma`` is a peak enhancement factor.
+    - ``alpha = 1 - 0.287 * ln(gamma)`` is a normalizing factor.
+    - ``sigma`` is the spectral width parameter:
+        - ``sigma = simga_a`` for ``w <= wp``
+        - ``sigma = sigma_b`` for ``w > wp``
+    - ``wp = 2pi / Tp`` is the angular spectral peak frequency.
```
Comment (ace-e4s) [inline src/waveresponse/_standardized1d.py:223]: `alpha` depends on `sigma`. If `alpha` is hard-coded but `sigma` is user-defined, then `Hs` is not preserved.

I think we should hard-code sigma values as well. If anyone needs to modify the common JONSWAP then they have to construct it themselves.

**Review by ace-e4s — CHANGES_REQUESTED** [review]

File: src/waveresponse/_standardized1d.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vegard-solum-4ss) [reply src/waveresponse/_standardized1d.py:30]: Fixed.

File: src/waveresponse/_standardized1d.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        as:
+
+        ``S(f) = 2*pi*S(w)``
+
+        where ``S(f)`` and ``S(w)`` are the same spectrum but expressed
+        in terms of Hz and rad/s, respectively.
+        """
+        args = (hs, tp, gamma, sigma_a, sigma_b)
+        alpha = self._alpha(*args)
+        b = self._b(*args)
```
[the remark this answers is not in the document]
Comment (vegard-solum-4ss) [reply src/waveresponse/_standardized1d.py:283]: Fixed.

File: src/waveresponse/_standardized1d.py (hunk 1 above)
Comment (vegard-solum-4ss) [reply src/waveresponse/_standardized1d.py:397]: Fixed.

File: docs/user_guide/standardized_spectra.rst (hunk 4 above)
Comment (vegard-solum-4ss) [reply docs/user_guide/standardized_spectra.rst:31]: Fixed.

**Review by ace-e4s — APPROVED** [review]
👍

**Pull request merged**